src/astarte-client/models/Token/index.ts

```diff
@@ -0,0 +1,80 @@
+/*
+   This file is part of Astarte.
+
+   Copyright 2020 Ispirata Srl
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+import jwt from 'jsonwebtoken';
+
+import type { AstarteJWT, AstarteDecodedJWT } from '../../types';
+
+const isTokenExpired = (decodedTokenObject: AstarteDecodedJWT) => {
+  if (decodedTokenObject.exp) {
+    const posix = Number.parseInt(decodedTokenObject.exp, 10);
+    const expiry = new Date(posix * 1000);
+    const now = new Date();
+
+    return expiry <= now;
+  }
+  return false;
+};
+
+const hasAstarteClaims = (decodedTokenObject: AstarteDecodedJWT) => {
+  // AppEngine API
+  if ('a_aea' in decodedTokenObject) {
+    return true;
+  }
+
+  // Realm Management API
+  if ('a_rma' in decodedTokenObject) {
+    return true;
+  }
+
+  // Pairing API
+  if ('a_pa' in decodedTokenObject) {
+    return true;
+  }
+
+  // Astarte Channels
+  if ('a_ch' in decodedTokenObject) {
+    return true;
+  }
+
+  return false;
+};
+
+type ValidationResult = 'expired' | 'notAnAstarteToken' | 'valid' | 'invalid';
+
+export const validateToken = (token: AstarteJWT): ValidationResult => {
+  const decoded = jwt.decode(token, { complete: true }) as AstarteDecodedJWT;
+
+  let status: ValidationResult = 'invalid';
+
+  if (decoded) {
+    if (isTokenExpired(decoded.payload)) {
```

Comment:
here you have the decoded token, but `isTokenExpired` decodes the token again.
If you can i'd leave that token change out of the PR, but if you can't it would be better to use a model for this token.
Decode once, and then query it's properties

Reply:
Indeed, I was trying to minimize changes and only add needed annotations for types. I am pushing a commit where I rework this file into an AstarteToken class.